fe/fe-core/src/main/java/com/starrocks/connector/hive/HiveConnectorScanRangeSource.java

```diff
@@ -0,0 +1,409 @@
+            hdfsScanRange.setFile_length(fileDesc.getLength());
+            hdfsScanRange.setModification_time(fileDesc.getModificationTime());
+
+            RemoteFileInputFormat fileFormat = remoteFileInfo.getFormat();
+            hdfsScanRange.setFile_format(fileFormat.toThrift());
+            if (fileFormat.isTextFormat()) {
+                hdfsScanRange.setText_file_desc(fileDesc.getTextFileFormatDesc().toThrift());
+            }
+
+            if (attachment.dataCacheOptions != null) {
+                TDataCacheOptions tDataCacheOptions = new TDataCacheOptions();
+                tDataCacheOptions.setPriority(attachment.dataCacheOptions.getPriority());
+                hdfsScanRange.setDatacache_options(tDataCacheOptions);
+            }
+
+            TScanRange scanRange = new TScanRange();
+            scanRange.setHdfs_scan_range(hdfsScanRange);
+            scanRangeLocations.setScan_range(scanRange);
+
+            if (blockDesc.isPresent()) {
+                if (blockDesc.get().getReplicaHostIds().length == 0) {
+                    String message = String.format("hdfs file block has no host. file = %s/%s",
+                            remoteFileInfo.getFullPath(), fileDesc.getFileName());
+                    throw new StarRocksPlannerException(message, ErrorType.INTERNAL_ERROR);
+                }
+
+                for (long hostId : blockDesc.get().getReplicaHostIds()) {
+                    String host = blockDesc.get().getDataNodeIp(hostId);
+                    TScanRangeLocation scanRangeLocation = new TScanRangeLocation(new TNetworkAddress(host, -1));
+                    scanRangeLocations.addToLocations(scanRangeLocation);
+                }
+            } else {
+                TScanRangeLocation scanRangeLocation = new TScanRangeLocation(new TNetworkAddress("-1", -1));
+                scanRangeLocations.addToLocations(scanRangeLocation);
+            }
+            buffer.add(scanRangeLocations);
+        }
+
+    }
+
+    public Iterator<TScanRangeLocations> createScanRangeIterator(RemoteFileInfo remoteFileInfo) {
+        return new ScanRangeIterator(remoteFileInfo);
+    }
+
+    private void updateBackendSplitFile(RemoteFileInfo partition) {
+        ConnectContext connectContext = ConnectContext.get();
+        if (connectContext == null) {
+            return;
+        }
+        long hugeFileSize = connectContext.getSessionVariable().getConnectorHugeFileSize();
+        for (RemoteFileDesc fileDesc : partition.getFiles()) {
+            if (fileDesc.isSplittable()) {
+                backendSplitCount += (fileDesc.getLength() + hugeFileSize - 1) / hugeFileSize;
+            } else {
+                backendSplitCount += 1;
+            }
+        }
+        // if splits is small comparing to nodes, then better not let backend do split.
+        int nodes = connectContext.getAliveComputeNumber() + connectContext.getAliveBackendNumber();
+        backendSplitFile = (backendSplitCount > 2 * nodes);
```

Comment:
Will backendSplitFile be modified as split accumulates?

Reply:
Yes,  if there are many splits already (exceeds BE nodes),  then we can let BE do split without worrying some BE does not get splits.

but if there is few splits, it's better to split file at FE side using small split size, so all BE nodes could get splits. 

In old code,  we iterate all files to check split number is large enough.  but in new code, since we are emitting splits incrementally, that's the best what we can do.